File: pageindicatorview/src/main/java/com/rd/draw/DrawManager.java
package com.rd.draw;

import android.content.Context;
import android.graphics.Canvas;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import com.rd.animation.data.Value;
import com.rd.draw.controller.AttributeController;
import com.rd.draw.controller.DrawController;
import com.rd.draw.controller.MeasureController;
import com.rd.draw.data.Indicator;

public class DrawManager {

    private Indicator indicator;
    private DrawController drawController;
    private MeasureController measureController;
    private AttributeController attributeController;

    public DrawManager() {
        this.indicator = new Indicator();
        this.drawController = new DrawController(indicator);
        this.measureController = new MeasureController();
        this.attributeController = new AttributeController(indicator);
    }

    @NonNull
    public Indicator indicator() {
        if (indicator == null) {
            indicator = new Indicator();
        }

        return indicator;
    }

    public void updateValue(@NonNull Value value) {
        drawController.updateValue(value);
    }

    public void draw(@NonNull Canvas canvas) {
        drawController.draw(canvas);
    }

    public int measureViewWidth(int widthMeasureSpec) {
        return measureController.measureViewWidth(indicator, widthMeasureSpec);
    }

    public int measureViewHeight(int heightMeasureSpec) {
        return measureController.measureViewHeight(indicator, heightMeasureSpec);
    }

    public void initAttributes(@NonNull Context context, @Nullable AttributeSet attrs) {
        attributeController.init(context, attrs);
    }
}

File: pageindicatorview/src/main/java/com/rd/draw/DrawManager.kt
package com.rd.draw

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import android.util.Pair
import android.view.MotionEvent
import com.rd.animation.data.Value
import com.rd.draw.controller.AttributeController
import com.rd.draw.controller.DrawController
import com.rd.draw.controller.MeasureController
import com.rd.draw.data.Indicator

class DrawManager {

	private val attributeController: AttributeController
	private val measureController: MeasureController
	private val drawController: DrawController
	private val indicator = Indicator()

	init {
		this.attributeController = AttributeController(indicator)
		this.drawController = DrawController(indicator)
		this.measureController = MeasureController()
	}

	fun indicator(): Indicator {
		return indicator
	}

	fun setClickListener(listener: DrawController.ClickListener?) {
		drawController.setClickListener(listener)
	}

	fun touch(event: MotionEvent?) {
		drawController.touch(event)
	}

	fun updateValue(value: Value?) {
		drawController.updateValue(value)
	}

	fun draw(canvas: Canvas) {
		drawController.draw(canvas)
	}

	fun measureViewSize(widthMeasureSpec: Int, heightMeasureSpec: Int): Pair<Int, Int> {
		return measureController.measureViewSize(indicator, widthMeasureSpec, heightMeasureSpec)
	}

	fun initAttributes(context: Context, attrs: AttributeSet?) {
		attributeController.init(context, attrs)
	}
}
